pallets/settlement/src/lib.rs

```diff
@@ -1118,22 +1269,119 @@ impl<T: Trait> Module<T> {
                 }
             }
 
+            enum FailureReason {
+                BaseTransfer,
+                ConfidentialTransfer,
+                UndefinedLeg,
+                InsufficientMercatAuthorizations,
+                GetMercatAccount,
+                GetMercatMediatorData,
+            }
+
             if !failed {
                 match with_transaction(|| {
                     for (leg_id, leg_details) in legs.iter().filter(|(leg_id, _)| {
                         let status = Self::instruction_leg_status(instruction_id, leg_id);
-                        status == LegStatus::ExecutionPending
+                        matches!(status, LegStatus::ExecutionPending)
                     }) {
-                        let result = T::Asset::base_transfer(
-                            leg_details.from,
-                            leg_details.to,
-                            &leg_details.asset,
-                            leg_details.amount,
-                        );
-                        if let Ok(post_info) = result {
-                            transaction_weight += post_info.actual_weight.unwrap_or_default();
-                        } else {
-                            return Err(leg_id);
+                        match leg_details {
+                            Leg::NonConfidentialLeg(leg) => {
+                                let result = T::Asset::base_transfer(
+                                    leg.from, leg.to, &leg.asset, leg.amount,
+                                );
+                                if let Ok(post_info) = result {
+                                    transaction_weight +=
+                                        post_info.actual_weight.unwrap_or_default();
+                                } else {
+                                    return Err((leg_id, FailureReason::BaseTransfer));
+                                }
+                            }
+                            Leg::ConfidentialLeg(leg) => {
+                                let mut tx_data = Self::mercat_tx_data(instruction_id);
+                                if tx_data.len() != 3 {
```

Comment:
This only holds when there is one confidential leg in the instruction. However, this is not enforced anywhere in the code. We should add a check to make sure only one confidential tx is allowed per instruction.

Reply:
Added a check during add instruction to ensure that there is only one confidential leg.